File: lib/dog/compiler/compiler.rb
#
# Copyright 2011 by Salman Ahmad (salman@salmanahmad.com).
# All rights reserved.
#
# Permission is granted for use, copying, modification, distribution,
# and distribution of modified versions of this work as long as the
# above copyright notice is included.
#

module Dog
  
  class CompilationError < RuntimeError
    attr_accessor :errors
  end
  
  class Compiler

    attr_accessor :bundle
    attr_accessor :errors

    def self.compile(units)
      compiler = self.new
      for unit in units do
        raise "A compilation units must contain two elements [AST, filename]" unless unit.size == 2
        compiler.compile(unit[0], unit[1])
      end

      compiler.finalize
    end

    def initialize
      self.bundle = Bundle.new
      self.errors = []
    end

    def compile(node, filename = "")
      if node && node.package then
        package_name = node.package
      else
        package_name = ""
      end

      package = self.bundle.packages[package_name]

      unless package then
        package = ::Dog::Package.new(package_name)
        self.bundle.link(package)
      end

      # TODO - Reintroduce the rule architecture
      #::Dog::Nodes::Node.each_descendant(node) do |d|
      #  rule = Rules::Rule.new(self)
      #  rule.apply(d)
      #end

      if errors.empty? && node then
        node.compile(package)
      end
    end

    def startup_package(package_name)
      self.bundle.startup_package = package_name
    end

    def finalize
      if errors.empty? then
        self.bundle.finalize
        self.bundle.sign
        return self.bundle
      else
        compilation_error = nil
        
        if errors.size == 1 then 
          failure_reason = "Compilation Error: There was #{errors.size} error that took place.\n\n#{errors.join("\n\n")}\n"
          compilation_error = CompilationError.new(failure_reason)
        else 
          failure_reason = "Compilation Error: There was #{errors.size} error that took place.\n\n#{errors.join("\n\n")}\n"
          compilation_error = CompilationError.new("Compilation Error: There were #{errors.size} errors that took place.\n\n#{errors.join("\n\n")}\n")
        end
          
        compilation_error.errors = errors
        raise compilation_error
      end
    end
    
    def report_error_for_node(node, description)
      self.errors << "(#{node.filename}:#{node.line}) - #{description}"
    end
    
  end
  
end

File: src/main/java/dog/lang/compiler/Compiler.java
/*
 *
 *  Copyright 2012 by Salman Ahmad (salman@salmanahmad.com).
 *  All rights reserved.
 *
 *  Permission is granted for use, copying, modification, distribution,
 *  and distribution of modified versions of this work as long as the
 *  above copyright notice is included.
 *
 */

package dog.lang.compiler;

import dog.lang.Bark;
import dog.lang.Resolver;
import dog.lang.nodes.Node;
import dog.lang.nodes.Nodes;
import dog.lang.nodes.Definition;
import dog.lang.nodes.FunctionDefinition;
import dog.lang.nodes.ConstantDefinition;
import dog.lang.nodes.StructureDefinition;

import org.apache.commons.lang3.StringUtils;

import java.util.ArrayList;

public class Compiler {
	ArrayList<Symbol> symbols = new ArrayList<Symbol>();
	Resolver resolver;
	Bark bark;

	public Compiler() {
		this(new Resolver());
	}

	public Compiler(Resolver r) {
		resolver = r;
	}

	public ArrayList<Symbol> getSymbols() {
		return symbols;
	}

	public Bark getBark() {
		return bark;
	}

	public Bark compile() {
		ArrayList<byte[]> bytecode = new ArrayList<byte[]>();

		for(Symbol symbol : symbols) {
			symbol.compile();
		}

		for(Symbol symbol : symbols) {
			bytecode.add(symbol.bytecode);
		}

		bark = new Bark(symbols.get(0).name, bytecode);
		return bark;
	}

	public void addCompilationUnit(Nodes ast, String sourceFileName) {
		ArrayList<String> packageName = ast.getPackageName();
		ArrayList<ArrayList<String>> includedPackages = ast.getIncludedPackages();
		ArrayList<ArrayList<String>> loadedPackages = ast.getLoadedPackages();

		boolean containsNonDefinitions = false;

		ArrayList<Node> nodes = Helper.descendantsOfNode(ast);

		for (Node node : nodes) {
			node.filePath = sourceFileName;
			node.packageName = packageName;
			node.includedPackages = includedPackages;
			node.loadedPackages = loadedPackages;
		}

		for(Node node : ast.children()) {
			if(!(node instanceof Definition)) {
				containsNonDefinitions = true;
			}
		}

		if(containsNonDefinitions) {
			Symbol root = new Function(StringUtils.join(packageName, ".") + "." + "@root", ast, this);
			this.addSymbol(root);
		}

		ast.scaffold(this);
	}

	public void addSymbol(Symbol symbol) {
		for(Symbol s : symbols) {
			if(s.name.equals(symbol.name)) {
				throw new RuntimeException("Duplicate symbol during compilation.");
			}
		}

		symbols.add(symbol);
	}

	public ArrayList<dog.lang.Symbol> searchForSymbolsStartingWith(String name) {
		ArrayList<dog.lang.Symbol> list = new ArrayList<dog.lang.Symbol>();
		for(Symbol symbol : symbols) {
			
			if(symbol.name.startsWith(name)) {

				if(symbol instanceof Constant) {
					list.add(new dog.lang.Symbol(symbol.name, dog.lang.Symbol.Kind.CONSTANT));
				}

				if(symbol instanceof Type) {
					list.add(new dog.lang.Symbol(symbol.name, dog.lang.Symbol.Kind.TYPE));
				}

				if(symbol instanceof Function) {
					list.add(new dog.lang.Symbol(symbol.name, dog.lang.Symbol.Kind.FUNCTION));
				}
			}
		}
		
		ArrayList<dog.lang.Symbol> resolvedList = resolver.searchForSymbolsStartingWith(name);
		list.addAll(resolvedList);

		return list;
	}

	public dog.lang.Symbol searchForSymbol(String name) {
		ArrayList<dog.lang.Symbol> symbols = searchForSymbolsStartingWith(name);
		for(dog.lang.Symbol symbol : symbols) {
			if(symbol.name.equals(name)) {
				return symbol;
			}
		}

		return null;
	}

}
